Patch for #1433 #1432 and test for wrong install of drivers.

GillesDuvert opened the pull request:
(edited): initially motivated by #1433, but also solve the problem with **test_usersym** and #1432

#1433 : hm, the algorithm to get the same automatic ranges as IDL is bit more… complicated than expected.

Indeed, we need to insure that the !X or !Y.CRANGE, computed by FLOOR and CEIL on the (double precision) [X|Y]RANGE values do not 'jump' to the next interval due to numeric precision problems. I note that IDL behaves differently in this respect when the  [X|Y]RANGE is explicitely set to double precision: b=randomu(seed, 100000) & offset=2399.999d & plot, b, xrange=[offset, offset+24*24-1] will set lower bound to 2200 b=randomu(seed, 100000) & offset=2399.999 & plot, b, xrange=[offset, offset+24*24-1] will set lower bound to 2400

also, found a bug for PLOT,[0] where X.CRANGE was [0,0] instead of [0,1]

GillesDuvert committed:
hm, the algorithm to get the same automatic ranges as IDL is bit more complicated than expected.
Indeed, we need to insure that the !X or !Y.CRANGE, computed by FLOOR and CEIL on the (double precision) [X|Y]RANGE values do not 'jump' to the next interval due to numeric precision problems.
I note that IDL behaves differently in this respect when the  [X|Y]RANGE is explicitely set to double precision:
b=randomu(seed, 100000) & offset=2399.999d & plot, b, xrange=[offset, offset+24*24-1] will set lower bound to 2200
b=randomu(seed, 100000) & offset=2399.999 & plot, b, xrange=[offset, offset+24*24-1] will set lower bound to 2400

also, found a bug for PLOT,[0] where X.CRANGE was [0,0] instead of [0,1]

GillesDuvert committed:
test_usersym passes again: a oops in the draw_polyline() function was preventing to plot single PSYM points.
So much for trying to speedup graphics!
I note that plplot makes plots faster with a small buffersize, to be investigated in a future attepmpt to incorporate plplot to gdl.

alaingdl commented:
Hi. Thanks to the correction.

On my side, before this PR, with the script, when I run "./scripts/build_gdl.sh test" this test not fails
But when I call it at the GDL> prompt or, in the testsuite dir with ../install/bin/gdl -e test_plot_ranges
no problem (no errors)
 
I do not see/understand why !
What is the trick ?

GillesDuvert commented:
> Hi. Thanks to the correction.
> 
> On my side, before this PR, with the script, when I run "./scripts/build_gdl.sh test" this test not fails But when I call it at the GDL> prompt or, in the testsuite dir with ../install/bin/gdl -e test_plot_ranges no problem (no errors)
> 
> I do not see/understand why ! What is the trick ?

I have seen many times cases where a test works when called directly, but fails when called in 'make test' (or "./scripts/build_gdl.sh test" that amounts to the same I hope). No clue..

.. Although several libraries, and especially the plplot drivers, must be installed at a specific location before use. Using "make test" one may use an installed driver that is not the one just compiled by "make".
This because GDL does not look for dependencies in relative (to the location of the executable) locations but in absolute ("/usr/local/share/gnadatalanguage..") locations.
This is annoying and I believe "build_gdl.sh test" should test the **installed** gdl (following  "build_gdl.sh install") and not the built gdl as it does today.

GillesDuvert committed:
oops a bad cut/paste resulting in #1432

GillesDuvert committed:
Add test_dynamic_driver.pro
This will fail if the 'mem' driver used by gdl is not our modified driver.
Meaning the build was not ok, probably because plplot is not linked with Dynamic Drivers options.

alaingdl commented:
I take some time to test and all seems to be OK now :smile: 

2 points : 
- you mention 
 `b=randomu(seed, 100000) & offset=2399.999d & plot, b, xrange=[offset, offset+2424-1]` will set lower bound to **2200** but I cannot reproduce that in IDL 8.8 & GDL  (both cases (f or d) yeld to 2000 -- 5000 . If I add /xstyle, the exact ranges are used, same in IDL & GDL (!X.crange :        2399.9990       4822.9990)

- concerning the tests in CI and the use of "our" driver, maybe it is just related to the fact that the (internal) path to the lib (*.so) are not well set through Cmake ?

GillesDuvert commented:
@alaingdl
yes, strange, I cannot reproduced it just now, as I wrote it.
But I have
```
IDL> plot, randomu(seed, 100000) , xrange=[2399.999, 2400+24*24-1] ;  -> [2400-3000]
IDL> plot, randomu(seed, 100000) , xrange=[2399.999d, 2400+24*24-1] ;-> [2200-3000]
```
go figure.
